numpy/f2py/crackfortran.py

```diff
@@ -2873,6 +2909,141 @@ def analyzevars(block):
+       (which is often incompatible with the original fortran indexing)
+    2) allow the parameter array to be accessed in python as a dictionary with
+       fortran indices as keys
+    We are choosing 2 for now.
+    """
+    dim_attrspecs = [item[9:] for item in varsn['attrspec']
+                     if item[:9] == 'dimension']
+    if len(dim_attrspecs) > 1:
+        raise ValueError('dimension may appear at most once.')
+    elif dim_attrspecs:
+        # This is an array parameter.
+        # First, we parse the dimension information
+        a = dim_attrspecs[0]
+        a = a.strip()
+        if a[0] != "(" or a[-1] != ")":
+            raise ValueError(f'dimension for {n} can\'t be parsed')
+        dimrange = a[1:-1]
+        dimrange = dimrange.split(',')
+        if len(dimrange) == 1:
+            # e.g. dimension(2) or dimension(-1:1)
+            dimrange = dimrange[0].split(':')
+            # now, dimrange is a list of 1 or 2 elements
+            if len(dimrange) == 1:
+                bound = param_parse(dimrange[0], params)
+                dimrange = range(1, int(bound)+1)
+            else:
+                lbound = param_parse(dimrange[0], params)
+                ubound = param_parse(dimrange[1], params)
+                dimrange = range(int(lbound), int(ubound)+1)
+        else:
+            outmess('param_eval: multidimensional array parameters not'
+                    ' supported: %s\n' % repr(n))
+
+        # Parse parameter value
+        v = v.strip('()').strip('/').split(',')
+        v_eval = []
+        for item in v:
+            try:
+                item = eval(item, g_params, params)
+            except Exception as msg:
+                v_eval.append(item)
+                outmess('param_eval: got "%s" on %s\n' % (msg, repr(item)))
+            v_eval.append(item)
+
+        p = dict([])
+        for i in range(len(dimrange)):
+            p[dimrange[i]] = v_eval[i]
+    else:
+        try:
+            p = eval(v, g_params, params)
+        except Exception as msg:
+            p = v
+            outmess('param_eval: got "%s" on %s\n' % (msg, repr(v)))
+    return p
+
+
+def param_parse(d, params):
+    """Recursively parse array dimensions.
+
+    Parses the declaration of an array variable or parameter
+    `dimension` keyword, and is called recursively if the
+    dimension for this array is a previously defined parameter
+    (found in `params`).
+
+    Parameters
+    ----------
+    d : str
+        Fortran expression describing the dimension of an array.
+    params : dict
+        Previously parsed parameters declared in the Fortran source file.
+
+    Returns
+    -------
+    out : str
+        Parsed dimension expression.
+
+    Examples
+    --------
+
+    * If the line being analyzed is
+
+      `integer, parameter, dimension(2) :: pa = (/ 3, 5 /)`
+
+      then `d = 2` and we return immediately, with
+
+      `param_parse(d, params) = '2'`
```

Comment:
Can you write this in the form of a doctest, with `>>>`?

Reply:
Let me know if this is what you had in mind. I could write a proper doctest but that would involve populating the entire `param` dictionary which I don't think is the idea. I tried looking for similar tests in the rest of the code but couldn't find any. Thanks again for your patience on this, I'm still learning! Feel free to keep commenting if you think anything can be improved.

Reply:
I've put a suggestion above - but yes, a minimally populated params dictionary is exactly what I'm looking for, since otherwise it's hard to tell what form that argument takes.